src/MizarData/FrameTrackManagerTest.cpp

```diff
@@ -0,0 +1,233 @@
+static const std::vector<FrameStartNs> kD3d9FrameStarts = {
+    FrameStartNs(10), FrameStartNs(20), FrameStartNs(40), FrameStartNs(100), FrameStartNs(200)};
+
+static std::vector<PresentEvent> MakePresentEvent(const std::vector<FrameStartNs>& starts) {
+  std::vector<PresentEvent> result;
+  std::transform(std::begin(starts), std::end(starts), std::back_inserter(result),
+                 [](const FrameStartNs start) {
+                   PresentEvent event;
+                   event.set_begin_timestamp_ns(*start);
+                   return event;
+                 });
+  return result;
+}
+
+static const std::vector<PresentEvent::Source> kEtwSources = {PresentEvent::kDxgi,
+                                                              PresentEvent::kD3d9};
+
+const absl::flat_hash_map<PresentEvent::Source, std::vector<FrameStartNs>> kEtwSourceToFrameStart =
+    MakeMap(kEtwSources,
+            std::vector<std::vector<FrameStartNs>>{kDxgiFrameStarts, kD3d9FrameStarts});
+const absl::flat_hash_map<PresentEvent::Source, std::vector<PresentEvent>>
+    kEtwSourceToPresentEvent = MakeMap(
+        kEtwSources, std::vector<std::vector<PresentEvent>>{MakePresentEvent(kDxgiFrameStarts),
+                                                            MakePresentEvent(kD3d9FrameStarts)});
+
+static std::pair<std::vector<ScopeInfo>, std::vector<PresentEvent::Source>> DecomposeSources(
+    const absl::flat_hash_map<FrameTrackId, FrameTrackInfo>& id_to_infos) {
+  std::vector<ScopeInfo> scope_id_infos;
+  std::vector<PresentEvent::Source> etw_sources;
+
+  for (const auto& [unused_id, info] : id_to_infos) {
+    Visit([&scope_id_infos](const ScopeInfo& info) { scope_id_infos.push_back(info); },
+          [&etw_sources](PresentEvent::Source source) { etw_sources.push_back(source); }, info);
+  }
+
+  return {scope_id_infos, etw_sources};
+}
+
+class ExpectFrameTracksHasFrameStartsForScopes : public ::testing::Test {
+ public:
+  ExpectFrameTracksHasFrameStartsForScopes() {
+    EXPECT_CALL(data_, GetCaptureData).WillRepeatedly(ReturnRef(capture_data_));
+  }
+
+  void SourceToPresentEventExpectCall() {
+    EXPECT_CALL(data_, source_to_present_events).WillRepeatedly(Return(kEtwSourceToPresentEvent));
+  }
+
+  void CaptureDataExpectCalls() {
+    EXPECT_CALL(capture_data_, GetAllProvidedScopeIds).WillRepeatedly(Return(kScopeIds));
+    EXPECT_CALL(capture_data_, GetScopeInfo).WillRepeatedly(Invoke([](const ScopeId id) {
+      return kScopeIdToInfo.at(id);
+    }));
+    EXPECT_CALL(capture_data_, GetTimersForScope)
+        .WillRepeatedly(Invoke([](const ScopeId id, uint64_t /*min*/, uint64_t /*max*/) {
+          return kScopeIdToSTimerInfoPtrs.at(id);
+        }));
+  }
+
+  void ExpectGetFrameStartsIsCorrect(FrameStartNs min_start, FrameStartNs max_start) {
+    for (const auto& [id, info] : frame_track_manager_.GetFrameTracks()) {
+      const std::vector<FrameStartNs> expected = Visit(
```

Comment:
`expected` is too vague here, please use a better name

Reply:
I agree, the name is vague. But what's wrong with the word "expected"? Expect as in `EXPECT_THAT`. I think, `ExpectGetFrameTracksReturnsExpectedValueForEachFrameTrack` is a better name.

Reply:
Oh, NVM, you mean the variable name, not the function name.

Reply:
`expected_frame_starts`?

Reply:
maybe a middle point?
expected_start_time_list?

Reply:
I don't exactly like the `list` suffix. Because it's a vector not a list. And I wouldn't like a `vector` suffix, because the actual container doesn't really matter. And the fact that it is a container is already reflected in the plural "starts".

Reply:
yes, sorry. I'm not having internet in the train. expected_frame_starts is perfect.

Reply:
No worries.